src/actions/restart.py

```diff
@@ -1,14 +1,18 @@
-import charms.contextual_status as status
 import ops
 from charms import kubernetes_snaps
 
+SERVICES = {
+    "api-server": "snap.kube-apiserver.daemon",
+    "controller-manager": "snap.kube-controller-manager.daemon",
+    "kube-scheduler": "snap.kube-scheduler.daemon",
+}
+
 
 def restart_action(event: ops.ActionEvent):
     """Handle the upgrade action."""
-    with status.context(event.framework.model.unit):
-        kubernetes_snaps.service_restart("snap.kube-apiserver.daemon")
-        event.set_results({"api-server": {"status": "restarted"}})
-        kubernetes_snaps.service_restart("snap.kube-controller-manager.daemon")
-        event.set_results({"controller-manager": {"status": "restarted"}})
-        kubernetes_snaps.service_restart("snap.kube-scheduler.daemon")
-        event.set_results({"kube-scheduler": {"status": "restarted"}})
+    try:
+        for service, snap in SERVICES.items():
+            kubernetes_snaps.service_restart(snap)
+            event.set_results({service: {"status": "restarted"}})
+    except Exception as e:
+        event.fail(str(e))
```

Comment:
@kwmonroe in case you missed it -- this runs this action in a try/catch rather than using `status.context` which would then require us to reconcile again.  I think this is a better approach to what the action intends